manifests/install.pp

```diff
@@ -2,10 +2,38 @@
 #
 # This class is called from puppet_agent for install.
 #
-class puppet_agent::install {
+# === Parameters
+#
+# [package_file_name]
+#   The puppet-agent package file name.
+#   (see puppet_agent::prepare::package_file_name)
+#
+class puppet_agent::install(
+  $package_file_name = undef,
+) {
   assert_private()
 
-  package { $::puppet_agent::package_name:
-    ensure => present,
+  if $::operatingsystem == 'SLES' and $::operatingsystemmajrelease == '10' {
+    if $puppet_agent::params::master_agent_version != $::aio_agent_version {
+
+      contain puppet_agent::install::remove_packages
+
+      exec { 'replace puppet.conf removed by package removal':
```

Comment:
Why do we need to uninstall everything? SLES 10 supports upgrading and dependency resolution.

```
b1o3ocw0433l686:~ # puppet --version
3.8.1 (Puppet Enterprise 3.8.2)
```
```
b1o3ocw0433l686:~ # rpm -qa | grep pe-*
openssl-certs-0.8.0-0.8.29
perl-Carp-Clan-5.3-13.5
perl-Digest-SHA1-2.10-15.2
perl-URI-1.35-15.2
limal-nfs-server-perl-1.1.72-0.3
openslp-server-1.2.0-22.34.1
perl-X500-DN-0.28-133.2
perl-Date-Calc-5.4-14.5
pe-augeas-1.3.0-1.pe.sles10
pe-mcollective-common-2.7.0.1-1.pe.sles10
pe-ruby-rgen-0.6.5-3.pe.sles10
pe-mcollective-2.7.0.1-1.pe.sles10
compat-openssl097g-0.9.7g-13.19.1
libjpeg-6.2.0-752.2
device-mapper-1.02.13-6.18.7
openct-0.6.6-16.4.1
perl-5.8.8-14.17.15
limal-perl-1.1.72-0.3
openssl-0.9.8a-18.48.17
perl-Compress-Zlib-1.35-14.2
perl-Config-Crontab-1.11-12.2
perl-Digest-MD4-1.5-13.2
perl-Parse-RecDescent-1.80-259.2
perl-TimeDate-1.16-136.2
perl-XML-Parser-2.34-43.2
perl-gettext-1.05-13.2
openslp-1.2.0-22.34.1
perl-Crypt-SmbHash-0.12-13.2
permissions-2007.2.15-0.7
limal-ca-mgm-perl-1.1.75-0.3
perl-Bootloader-0.4.19.24-0.6.33
suspend-20071205-0.11.13
OpenIPMI-2.0.14-2.7.12
yast2-perl-bindings-2.13.11-0.22
zypper-0.6.201-0.4.1
pe-puppet-enterprise-release-3.8.2.0-1.pe.sles10
pe-libldap-2.4.39-5.pe.sles10
pe-libyaml-0.1.6-5.sles10
pe-ruby-stomp-1.3.3-1.pe.sles10
pe-ruby-augeas-0.5.0-7.pe.sles10
pe-hiera-1.3.4.5-1.pe.sles10
pe-ruby-shadow-2.2.0-4.pe.sles10
pe-facter-2.4.4.0-1.pe.sles10
pe-puppet-3.8.1.1.50-1.pe.sles10
freetype2-2.1.10-18.23.1
perl-Config-IniFiles-2.39-13.4
perl-TermReadKey-2.30-13.2
perl-XML-Writer-0.600-13.2
openldap2-client-2.3.32-0.37.1
perl-Bit-Vector-6.4-13.5
opensc-0.9.6-17.12
openssh-5.1p1-41.8.20
pe-openssl-1.0.0s-1.pe.sles10
pe-ruby-1.9.3.551-2.pe.sles10
pe-rubygem-deep-merge-1.0.0-6.pe.sles10
pe-virt-what-1.14-2.sles10
pe-ruby-ldap-0.9.12-7.pe.sles10
```
```
b1o3ocw0433l686:~ # md5sum /etc/puppetlabs/puppet/puppet.conf
cab5c77b74732923b5830b9f4a1bdcab  /etc/puppetlabs/puppet/puppet.conf
b1o3ocw0433l686:~ # cat !$
cat /etc/puppetlabs/puppet/puppet.conf
[main]
    vardir = /var/opt/lib/pe-puppet
    logdir = /var/log/pe-puppet
    rundir = /var/run/pe-puppet
    basemodulepath = /etc/puppetlabs/puppet/modules:/opt/puppet/share/puppet/modules
    user  = pe-puppet
    group = pe-puppet
    archive_files = true
server = pe38-latest

[agent]
    report = true
    classfile = $vardir/classes.txt
    localconfig = $vardir/localconfig
    graph = true
    pluginsync = true
certname = b1o3ocw0433l686.delivery.puppetlabs.net
```
```
b1o3ocw0433l686:~ # rpm -U --replacepkgs puppet-agent-1.2.5.rpm
warning: puppet-agent-1.2.5.rpm: Header V3 RSA/SHA1 signature: NOKEY, key ID 4bd6ec30
warning: /etc/puppetlabs/mcollective/server.cfg created as /etc/puppetlabs/mcollective/server.cfg.rpmnew
warning: /etc/puppetlabs/puppet/puppet.conf created as /etc/puppetlabs/puppet/puppet.conf.rpmnew
insserv: script puppet: service puppet already provided!
insserv: script pe-mcollective: service mcollective already provided!
pe-puppet                 0:off  1:off  2:off  3:off  4:on   5:off  6:off
pe-mcollective            0:off  1:off  2:on   3:on   4:on   5:on   6:off
```
```
b1o3ocw0433l686:~ # /opt/puppetlabs/bin/puppet --version
4.2.2
```
```
b1o3ocw0433l686:~ # rpm -qa | grep pe-*
openssl-certs-0.8.0-0.8.29
perl-Carp-Clan-5.3-13.5
perl-Digest-SHA1-2.10-15.2
perl-URI-1.35-15.2
limal-nfs-server-perl-1.1.72-0.3
openslp-server-1.2.0-22.34.1
perl-X500-DN-0.28-133.2
perl-Date-Calc-5.4-14.5
compat-openssl097g-0.9.7g-13.19.1
libjpeg-6.2.0-752.2
device-mapper-1.02.13-6.18.7
openct-0.6.6-16.4.1
perl-5.8.8-14.17.15
limal-perl-1.1.72-0.3
openssl-0.9.8a-18.48.17
perl-Compress-Zlib-1.35-14.2
perl-Config-Crontab-1.11-12.2
perl-Digest-MD4-1.5-13.2
perl-Parse-RecDescent-1.80-259.2
perl-TimeDate-1.16-136.2
perl-XML-Parser-2.34-43.2
perl-gettext-1.05-13.2
openslp-1.2.0-22.34.1
perl-Crypt-SmbHash-0.12-13.2
permissions-2007.2.15-0.7
limal-ca-mgm-perl-1.1.75-0.3
perl-Bootloader-0.4.19.24-0.6.33
suspend-20071205-0.11.13
OpenIPMI-2.0.14-2.7.12
yast2-perl-bindings-2.13.11-0.22
zypper-0.6.201-0.4.1
puppet-agent-1.2.5-1.sles10
freetype2-2.1.10-18.23.1
perl-Config-IniFiles-2.39-13.4
perl-TermReadKey-2.30-13.2
perl-XML-Writer-0.600-13.2
openldap2-client-2.3.32-0.37.1
perl-Bit-Vector-6.4-13.5
opensc-0.9.6-17.12
openssh-5.1p1-41.8.20
```
```
b1o3ocw0433l686:~ # md5sum /etc/puppetlabs/puppet/puppet.conf
cab5c77b74732923b5830b9f4a1bdcab  /etc/puppetlabs/puppet/puppet.conf
b1o3ocw0433l686:~ # cat /etc/puppetlabs/puppet/puppet.conf
[main]
    vardir = /var/opt/lib/pe-puppet
    logdir = /var/log/pe-puppet
    rundir = /var/run/pe-puppet
    basemodulepath = /etc/puppetlabs/puppet/modules:/opt/puppet/share/puppet/modules
    user  = pe-puppet
    group = pe-puppet
    archive_files = true
server = pe38-latest

[agent]
    report = true
    classfile = $vardir/classes.txt
    localconfig = $vardir/localconfig
    graph = true
    pluginsync = true
certname = b1o3ocw0433l686.delivery.puppetlabs.net
```